src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlRetry/SqlRetryService.cs

```diff
@@ -383,5 +383,102 @@ public async Task TryLogEvent(string process, string status, string text, DateTi
                 // do nothing;
             }
         }
+
+        private static void InitReplicaHandler()
+        {
+            if (_replicaHandler == null) // this is needed, strictly speaking, only if SqlRetryService is not singleton, but it works either way.
+            {
+                lock (_initLocker)
+                {
+                    _replicaHandler ??= new ReplicaHandler();
+                }
+            }
+        }
+
+        private class ReplicaHandler
+        {
+            private DateTime? _lastUpdated;
+            private object _databaseAccessLocker = new object();
+            private double _replicaTrafficRatio = 0;
+            private long _usageCounter = 0;
+
+            public ReplicaHandler()
+            {
+            }
+
+            public async Task<SqlConnection> GetConnection(ISqlConnectionBuilder sqlConnectionBuilder, bool isReadOnly, CancellationToken cancel)
+            {
+                SqlConnection conn;
+                if (!isReadOnly)
+                {
+                    conn = await sqlConnectionBuilder.GetSqlConnectionAsync(initialCatalog: null, cancellationToken: cancel).ConfigureAwait(false);
+                }
+                else
+                {
+                    var replicaTrafficRatio = GetReplicaTrafficRatio(sqlConnectionBuilder);
+
+                    if (replicaTrafficRatio == 0)
+                    {
+                        conn = await sqlConnectionBuilder.GetSqlConnectionAsync(initialCatalog: null, cancellationToken: cancel).ConfigureAwait(false);
+                    }
+                    else if (replicaTrafficRatio == 1)
+                    {
+                        conn = await sqlConnectionBuilder.GetReadOnlySqlConnectionAsync(initialCatalog: null, cancellationToken: cancel).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        var useWriteConnection = Interlocked.Increment(ref _usageCounter) % (int)(1 / (1 - _replicaTrafficRatio)) == 1; // examples for ratio -> % divider = { 0.9 -> 10, 0.8 -> 5, 0.75 - 4, 0.67 - 3, 0.5 -> 2, <0.5 -> 1}
```

Comment:
Should we consider: 

```suggestion
                        var useWriteConnection = unchecked(Interlocked.Increment(ref _usageCounter)) % (int)(1 / (1 - _replicaTrafficRatio)) == 1; // examples for ratio -> % divider = { 0.9 -> 10, 0.8 -> 5, 0.75 - 4, 0.67 - 3, 0.5 -> 2, <0.5 -> 1}
```
Assuming we might not overflow with long, but this should reset to 0 if we do.

Reply:
Will do.